snapshot.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	return false, nil
+}
+
+// encodeFileSnapshot encodes a snapshot file into a protobuf object.
+func encodeFileSnapshot(f *fileSnapshot) *internal.FileSnapshot {
+	return &internal.FileSnapshot{
+		Name:    proto.String(f.name),
+		Hash:    proto.Uint64(f.hash),
+		Content: proto.String(f.content),
+	}
+}
+
+// decodeFileSnapshot decodes a snapshot file from a protobuf object.
+func decodeFileSnapshot(pb *internal.FileSnapshot) *fileSnapshot {
+	return &fileSnapshot{
+		name:    pb.GetName(),
+		hash:    pb.GetHash(),
+		content: pb.GetContent(),
+	}
+}
+
+// fileSnapshots represents a list of snapshot files.
+type fileSnapshots []*fileSnapshot
+
+// newFileSnapshots returns a slice of stat'd snapshot files.
+func newFileSnapshots(path string, names []string) ([]*fileSnapshot, error) {
+	// Sort filenames for consistency.
+	sort.Strings(names)
+
+	// Build list of snapshot files with current stats.
+	// Ignore files that have been deleted. They are likely temporary files.
+	a := make([]*fileSnapshot, 0, len(names))
+	for _, name := range names {
+		f, err := newFileSnapshot(path, name)
+		if os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		a = append(a, f)
+	}
+	return a, nil
+}
+
+// isDirty returns true after finding the first dirty file.
+func (a fileSnapshots) isDirty(path string) (bool, error) {
+	for _, f := range a {
+		v, err := f.isDirty(path)
+		if err != nil {
+			return false, err
+		} else if v {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// encodeFileSnapshots encodes a slice of snapshot files into a protobuf object.
+func encodeFileSnapshots(a []*fileSnapshot) []*internal.FileSnapshot {
+	pb := make([]*internal.FileSnapshot, len(a))
+	for i := range a {
+		pb[i] = encodeFileSnapshot(a[i])
+	}
+	return pb
+}
+
+// decodeFileSnapshots decodes a slice of snapshot files from a protobuf object.
+func decodeFileSnapshots(pb []*internal.FileSnapshot) []*fileSnapshot {
+	a := make([]*fileSnapshot, len(pb))
+	for i := range pb {
+		a[i] = decodeFileSnapshot(pb[i])
+	}
+	return a
+}
+
+// hashTarget returns a hash for a target based on its commands and dependencies.
+func hashTarget(t *Target) uint64 {
+	h := fnv.New64a()
+	hashWriteStrings(h, t.Dependencies)
+
+	for _, c := range t.Commands {
+		switch c := c.(type) {
+		case *ExecCommand:
+			h.Write([]byte("exec"))
+			hashWriteStrings(h, c.Args)
+		case *ShellCommand:
+			h.Write([]byte("shell"))
+			h.Write([]byte(c.Source))
+		default:
+			panic("unreachable")
+		}
+	}
+
+	return h.Sum64()
+}
+
+// hashFileInfo generates a hash for a file or directory info.
+// Regular files are hashed using their mode, mtime & size.
+// Directories are hashed using its mode and a list of the directory's files.
+func hashFileInfo(path string) (uint64, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return 0, err
+	}
+
+	// If it's a regular file then hash the mode, mtime, & size.
+	if !fi.IsDir() {
+		h := fnv.New64a()
```

Comment:
I think we should be using SHA-256 everywhere, no reason to risk collisions with a tiny non-cryptographic hash.